docs/content/sips/008-using-oci-registries.md

```diff
@@ -0,0 +1,298 @@
+
+```bash
+$ spin oci login --username <username> --password <password>
+# OR
+$ echo $CONTAINER_REGISTRY_PASSWORD | spin oci login --username <username> --password-stdin
+```
+
+This user experience would mirror
+[the `docker login` command](https://docs.docker.com/engine/reference/commandline/login/).
+
+### Migrating applications from Bindle to an OCI registry
+
+Changing the distribution mechanism for Spin applications is a breaking change
+for the project - to address this, the change should be implemented gradually,
+giving users the tool to migrate an application from Bindle to an OCI registry.
+
+This would be done in two steps: pulling the application from Bindle, then
+pushing it to the desired OCI registry:
+
+```bash
+$ spin bindle pull --bindle-server <server> --bindle-username <username> --bindle-password <password> [bindle-name]
+# then
+$ spin oci push
+```
+
+### Implementation
+
+A Spin application is made up of some metadata and component information
+together with the Wasm modules and static assets that made up those components.
+So conceptually, a Spin application is not _a single artifact_, but rather
+multiple distinct objects.
+
+This SIP proposes that, when distributed using an OCI registry, a Spin
+application would become a new OCI _artifact_ with multiple layers
+(making the distinction clear, as _images_ and _artifacts_ are separate
+entities in OCI). Specifically, the media type used in this implementation
+for a Spin application is `application/vnd.fermyon.spin.application.v1+config`,
+and each Wasm module and static asset from the Spin components becomes an individual
+_layer_ in the resulting registry entity. Because each file and Wasm module becomes
+a separate _layer_, we can efficiently de-duplicate and distribute applications.
+
+The remaining question is representing the Spin application definition - Spin
+introduced the internal representation of a _locked application_ - an
+intermediate representation of a Spin application that _can_ have a way to
+content-address Wasm modules and static assets. The implementation for this SIP
+uses the Spin locked application as the OCI configuration object.
+
+Below is an example of a Spin application with one component and one static
+asset, its OCI manifest, locked application, and resulting local directory
+structure when pulling the application locally.
+
+Consider the following `spin.toml` file for the application:
+
+```toml
+spin_version = "1"
+authors = ["Radu Matei <radu.matei@fermyon.com>"]
+description = ""
+name = "github-stars-webhook"
+trigger = { type = "http", base = "/" }
+version = "0.1.0"
+
+[[component]]
+id = "github-star-webhook"
+source = "target/spin-http-js.wasm"
+files = ["my-file.json"]
+allowed_http_hosts = ["https://hooks.slack.com"]
+[component.trigger]
+route = "/..."
+```
+
+Distributing it to the GitHub Container Registry:
+
+```bash
+$ spin oci push ghcr.io/radu-matei/spin-example:v1
```

Comment:
I notice here you're passing an explicit version.  Is that how you see this working in production?

Versioning does seem like an area we should define carefully (or at least identify as requiring mindful discussion during implementation).  Immutability was both a benefit and a curse of Bindle, and we ended up using bonus versions (content or random) as a way to update the bindle store without revving the Spin app version.  Will we derive the OCI version from the Spin app version?  Will that result in mutation?  What consequences will that have?

(I realise the consequences may not be seen until we look at the Platform/Cloud side.  But by that point we will already have established how versions are set/used.  So I do reckon we need to think about it now.)

Reply:
This is the first can of worms of using OCI for this.
I debated whether to go on a tangent in the SIP, but decided against it...

As you point out in the comment, tags in Bindle were immutable, whereas tags in OCI are famously mutable (see `latest`...).


The first prototype for this had the following behavior: 
- if `spin oci push` had no arguments, the full OCI reference would be derived from `<app-name>:v<version>`
- if `spin oci push` had 1 argument, the OCI reference, that would be used instead.

For `spin oci push`, which would be the method for distributing applications, decoupled from Platform/Cloud, users being able to control the tag independently is something I would like to have.

Platform/Cloud represents an opinionated way to distribute and deploy the application. The current behavior of using the build metadata to push a new tag makes sense as the starting point IMO.

What do you think?

Reply:
You should _totally_ go on that 'tangent' - it's really useful for a SIP to discuss alternatives and trade-offs, and explain why a particular design choice was preferred.  Otherwise we hit a situation where we are on the wrong side of the chosen trade-off and we have to rediscover why it's better than the alternatives, usually accompanied by much grumbling!  And I do think this is an important and subtle design point, not a tangent at all!

Maybe I'm focused too much on interactive scenarios, but I do like some ability to derive the name and version from the manifest, if for no other reason than to avoid typos.  People are just not going to get it right every time.

But mutability does make that dangerous.  You don't want to do a push and find you've blatted a stable tag with an experimental release because you forgot to bump the version.  We could resume the Bindle strategy of appending a bonus sub-version, and require users to opt out by passing the `--no-build-metadata` (really `--this-really-is-the-real-version-for-real`) flag.  But then maybe we're trying to recreate Bindleisms on top of a fundamentally different model.  Maybe we embrace mutability, and if consumers want stability they can reference by digest instead of version.

(There's also a reasonable argument that "well really you should only be publishing from CI, we intentionally made it so you have to be more deliberate when pushing random code from a random laptop," but again, if that's a factor then we should capture it.)

In short, I am conflicted.  Sorry for the ramble, I am very much thinking out loud!

Could you say more about why you decided against your first approach?  It seems like the 1-arg version would have allowed "users ... to control the tag independently" but not _required_ them to manage it on every single interaction.

(Oh, and re Platform/Cloud/deploy, yeah you're right, in a sense the whole versioning story is kind of an internal implementation detail, at least for now.  Though it will become significant again if we bring back the auto-rev channels from the original Hippo.)

Reply:
Fundamentally, `spin oci push` and `spin deploy` are two entirely different operations, that target different use cases — which is the reason I thought expanding in this SIP for the changes to `spin deploy` would have been unnecessary (or a 'tangent'), as I think it would require its own SIP.
However, you are right, expanding in this SIP is necessary, and I'd like to use this conversation to inform the SIP before updating it.


Let me expand a bit on why I think they are different operations, even if they happen to (and really _should_) use the same distribution mechanism.

`spin oci push` is intended to give a user the option to distribute their application using widely available registry services — and to be paired with `spin up`. The idea behind this pair of operations is to give a user as much flexibility and freedom in choosing their strategy for _how_ to do versioning, tagging, and to make their own decision in relation to tag mutability.
Some people really _do_ want and _need_ to use `:latest`, despite what opinions we might or might not have on the subject.
Deriving the name and tag from `spin.toml` when no explicit name and tag are defined is something we can do — however, this would imply two things:

- the name of the application now has to reference the registry URL, account, and repository — otherwise, trying to push to the registry/user/repository `my-simple-app` makes no sense.
- if the tag is derived from the version, we are restricting tags to semantic versioning, something that OCI does not dictate.

To sum up this up, having _the option_ to derive the name and tag from `spin.toml` sounds good — even though I might consider having a flag for this instead of it being the default (I don't have too strong of an opinion here though).

Reply:
That sounds good to me, and we can iterate on whether "derive name and version from manifest" should be default or opt-in - it doesn't affect the design in significant ways.  I am sure folks like Mikkel will have a view on this.  If we can capture this discussion into the SIP then I'm happy with the conclusion we've reached.  Thanks!

Reply:
(even if you want to capture by reference by linking back to the PR here... just something so that when we look at the SIP we can easily find the discussion)

Reply:
Tried to capture the context for this in the SIP.
Let me know if you think it does capture it or you think that section can be further expaded.

Thanks, the conversation so far on this has been really helpful!